### PR TITLE
AnnotationColumn struct to fix hard tab column numbers in errors

### LINKED ISSUE #109537: Inconsistent column numbers when using hard tabs
### Code

src/main.rs:
```Rust
mod m;
fn main() {
	let s = m::S;
	s.method();
}
```

src/m.rs:
```Rust
pub struct S;
impl S {
		fn method(&self) {}
}
```

### Current output

```Shell
error[E0624]: associated function `method` is private
 --> src/main.rs:4:4
  |
4 |     s.method();
  |       ^^^^^^ private associated function
  |
 ::: src/m.rs:3:9
  |
3 |         fn method(&self) {}
  |         ---------------- private associated function defined here
```

### Desired output

```Shell
error[E0624]: associated function `method` is private
 --> src/main.rs:4:4
  |
4 |     s.method();
  |       ^^^^^^ private associated function
  |
 ::: src/m.rs:3:3
  |
3 |         fn method(&self) {}
  |         ---------------- private associated function defined here
```

### Rationale and extra context

for the error in `main.rs`, rust gives the column number as a "character index" (i.e. the error is located at the 4th character in the line) which i  think is the intended behavior, but for the reference to `m.rs`, it seems to be treating a tab as 4 columns

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by pommicket**
Fixes #109537

i don't know if this is the best way of fixing this but it works

### CODE REVIEW

**Comment by petrochenkov** [comment]
>for the error in `main.rs`, rust gives the column number as a "character index" (i.e. the error is located at the 4th character in the line) which i  think is the intended behavior, but for the reference to `m.rs`, it seems to be treating a tab as 4 columns

Why is the right number reported in `main.rs`?
The `file` value wasn't previously kept and the `display` value should be `7` for `main.rs`, but the compiler still managed report the correct value somehow. What is the difference between `main.rs` and `m.rs`?

@rustbot author

**Comment by pommicket** [comment]
There are multiple places in the code where error line/column references are generated:

So-called "primary" messages are generated here (emitter.rs:1504 in method `EmitterWriter::emit_message_default`) — this is where the reference to `main.rs` is generated:

```rust
if !self.short_message {
    // remember where we are in the output buffer for easy reference
    let buffer_msg_line_offset = buffer.num_lines();

    buffer.prepend(buffer_msg_line_offset, "--> ", Style::LineNumber);
    buffer.append(
        buffer_msg_line_offset,
        &format!(
            "{}:{}:{}",
            sm.filename_for_diagnostics(&loc.file.name),
            sm.doctest_offset_line(&loc.file.name, loc.line),
            loc.col.0 + 1,
        ),
        Style::LineAndColumn,
    );
    for _ in 0..max_line_num_len {
        buffer.prepend(buffer_msg_line_offset, " ", Style::NoStyle);
    }
} else {
    buffer.prepend(
        0,
        &format!(
            "{}:{}:{}: ",
            sm.filename_for_diagnostics(&loc.file.name),
            sm.doctest_offset_line(&loc.file.name, loc.line),
            loc.col.0 + 1,
        ),
        Style::LineAndColumn,
    );
}
```
Here we have access to the `Loc` which includes the correct file column number.

Whereas the `m.rs` reference comes from here (line 1547):
```rust
let loc = if let Some(first_line) = annotated_file.lines.first() {
    let col = if let Some(first_annotation) = first_line.annotations.first() {
        format!(":{}", first_annotation.start_col + 1)
    } else {
        String::new()
    };
    format!(
        "{}:{}{}",
        sm.filename_for_diagnostics(&annotated_file.file.name),
        sm.doctest_offset_line(&annotated_file.file.name, first_line.line_index),
        col
    )
} else {
    format!("{}", sm.filename_for_diagnostics(&annotated_file.file.name))
};
```
Here we only have an `Annotation`, so its `start_col` member was used even though it actually represents the display column.

Another possibility for fixing this (other than what I've done in my branch) is to somehow recover/keep the column number from `msp.span_labels` but I'm not sure if it would be any easier or better.

File: compiler/rustc_errors/src/snippet.rs [hunk 1]
```diff
@@ -1,20 +1,46 @@
...
+    /// ```
+    ///
+    /// we want to keep around the correct file offset so that column numbers in error messages
+    /// are correct. (motivated by <https://github.com/rust-lang/rust/issues/109537>)
+    pub file: usize,
+}
+
+impl AnnotationColumn {
+    pub fn from_loc(loc: &Loc) -> Self {
+        Self { display: loc.col_display, file: loc.col.0 }
```
Comment (petrochenkov) [inline compiler/rustc_errors/src/snippet.rs:33]:
```suggestion
        AnnotationColumn { display: loc.col_display, file: loc.col.0 }
```
Nit: could you avoid using `Self` in non-generic contexts?

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**